apollo-federation/src/connectors/json_selection/apply_to.rs

```diff
@@ -3346,6 +3689,143 @@ mod tests {
+                }
+            }
+        });
+
+        let (result, errors) = JSONSelection::parse("$.user.profile ?{ name email }")
+            .unwrap()
+            .apply_to(&data);
+        assert_eq!(
+            result,
+            Some(json!({
+                "name": "Alice",
+                "email": "alice@example.com"
+            }))
+        );
+        assert_eq!(errors, vec![]);
+
+        // Test with null nested data
+        let data_with_null_profile = json!({
+            "user": {
+                "profile": null
+            }
+        });
+
+        let (result, errors) = JSONSelection::parse("$.user.profile ?{ name email }")
+            .unwrap()
+            .apply_to(&data_with_null_profile);
+        assert_eq!(result, None);
+        assert_eq!(errors, vec![]);
+    }
+
+    #[test]
+    fn test_mixed_optional_selection_and_optional_chaining() {
+        use serde_json_bytes::json;
+
+        let data = json!({
+            "user": {
+                "id": 123,
+                "profile": null
+            }
+        });
+
+        let (result, errors) = JSONSelection::parse("$.user ?{ id profileName: profile?.name }")
+            .unwrap()
+            .apply_to(&data);
+        assert_eq!(
+            result,
+            Some(json!({
+                "id": 123
+            }))
+        );
+        assert_eq!(errors, vec![]);
+
+        // Test with missing user
+        let data_no_user = json!({
+            "other": "value"
+        });
+
+        let (result, errors) = JSONSelection::parse("$.user ?{ id profileName: profile?.name }")
+            .unwrap()
+            .apply_to(&data_no_user);
+        assert_eq!(result, None);
+        assert_eq!(errors.len(), 0);
+    }
+
+    #[test]
+    fn test_optional_selection_set_parsing() {
+        // Test that the parser correctly handles optional selection sets
+        let selection = JSONSelection::parse("$.user? { id name }").unwrap();
+        assert_eq!(selection.pretty_print(), "$.user? { id name }");
+
+        // Test with nested optional selection sets
+        let selection = JSONSelection::parse("$.user.profile? { name }").unwrap();
+        assert_eq!(selection.pretty_print(), "$.user.profile? { name }");
+
+        // Test mixed with regular selection sets
+        let selection = JSONSelection::parse("$.user? { id profile { name } }").unwrap();
+        assert_eq!(selection.pretty_print(), "$.user? { id profile { name } }");
+    }
+
+    #[test]
+    fn test_optional_selection_set_with_arrays() {
+        use serde_json_bytes::json;
+
+        let data = json!({
+            "users": [
+                {
+                    "id": 1,
+                    "name": "Alice"
+                },
+                null,
+                {
+                    "id": 3,
+                    "name": "Charlie"
+                }
+            ]
+        });
+
+        let (result, errors) = JSONSelection::parse("$.users ?{ id name }")
```

Comment:
In my follow-up PR #7993, I will most likely change these tests to use `#[rstest]` and `JSONSelection::parse_with_spec`, and version-gate the `?` syntax to `connect/v0.3` only (open to discussion/objections, but we're also version-gating `...` to v0.3, fwiw).

No need to address that here, but if we cut a preview release with just your current PR, then `?` will temporarily be available regardless of `ConnectSpec` version (🙀).